pkg/controller/build/build_controller.go

```diff
@@ -218,3 +244,91 @@ func (r *ReconcileBuild) validateOutputSecret(n string, ns string) error {
 	}
 	return nil
 }
+
+func (r *ReconcileBuild) configFinalizer(b *build.Build) error {
+	if b.Annotations != nil && b.Annotations[build.AnnotationBuildRunDeletion] == "true" {
+		if !contains(b.GetFinalizers(), build.BuildFinalizer) {
+			b.SetFinalizers(append(b.GetFinalizers(), build.BuildFinalizer))
+			err := r.client.Update(context.TODO(), b)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		if contains(b.GetFinalizers(), build.BuildFinalizer) {
+			b.SetFinalizers(remove(b.GetFinalizers(), build.BuildFinalizer))
+			err := r.client.Update(context.TODO(), b)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileBuild) finalizeBuildRun(b *build.Build) (bool, error) {
+	// Check if the build is marked to be deleted, which is
+	// indicated by the deletion timestamp being set.
+	if !b.DeletionTimestamp.IsZero() {
+		if contains(b.GetFinalizers(), build.BuildFinalizer) {
```

Comment:
Im not sure why this needs to be done again, at this point you already know that if `AnnotationBuildRunDeletion` is true, then a BuildRun should be deleted, can you assert for the annotation value instead?

Reply:
We need to make sure the build is at delete status by using this `DeletionTimestamp`, then start deleting the buildRun.
We cannot delete the buildRun when the annotations exist, it is just a config

Reply:
so I think `finalizeBuildRun` as well as `configFinalizer` should be called differently, you currently call one after the other, I think you should:

- only call `finalizeBuildRun` inside the Reconcile if a Build is being deleted, and this needs to happen inmediately after you get the build object, and before you try to add a new finalizer.

- only call `configFinalizer` after your `Build` passed all validations, because if the `Build` is not passing all validations, you will not be able to move forward, in other words, to create a `BuildRun`.

Reply:
- Yes, `finalizeBuildRun` only be called when a Build is being deleted. I just put the condition check in the finalizeBuildRun method: `if !b.DeletionTimestamp.IsZero() {`, but I agree it is better if we move the condtion out of the method, it is more clear, changed to:
```
	// Check if the build is marked to be deleted, which is
	// indicated by the deletion timestamp being set.
	if !b.DeletionTimestamp.IsZero() {
		finishClean, err := r.finalizeBuildRun(b)
		if err != nil {
			return reconcile.Result{}, err
		}
		if finishClean {
			return reconcile.Result{}, nil
		}
	}
```
- For the `configFinalizer`, I think it is fine, because this method is just to use to add the Finalizer in the Build, it is not related with any buildRun, no matter is there is no buildrun or with many buildruns. So it can be anywhere of the Reconciliation logic.